asyncstdlib/functools.py

```diff
@@ -45,44 +50,172 @@ def cache(user_function: AC) -> LRUAsyncCallable[AC]:
     return lru_cache(maxsize=None)(user_function)
 
 
-class AwaitableValue(Generic[T]):
+class AwaitableValue(Generic[R]):
     """Helper to provide an arbitrary value in ``await``"""
 
     __slots__ = ("value",)
 
-    def __init__(self, value: T):
+    def __init__(self, value: R):
         self.value = value
 
     # noinspection PyUnreachableCode
-    def __await__(self) -> Generator[None, None, T]:
+    def __await__(self) -> Generator[None, None, R]:
         return self.value
         yield  # type: ignore # pragma: no cover
 
     def __repr__(self) -> str:
         return f"{self.__class__.__name__}({self.value!r})"
 
 
-class _RepeatableCoroutine(Generic[T]):
-    """Helper to ``await`` a coroutine also more or less than just once"""
+class _FutureCachedValue(Generic[R, T]):
+    """A placeholder object to control concurrent access to a cached awaitable value.
+
+    When given a lock to coordinate access, only the first task to await on a
+    cached property triggers the underlying coroutine. Once a value has been
+    produced, all tasks are unblocked and given the same, single value.
+
+    """
 
-    __slots__ = ("call", "args", "kwargs")
+    __slots__ = ("_get_attribute", "_instance", "_name", "_lock")
 
     def __init__(
-        self, __call: Callable[..., Coroutine[Any, Any, T]], *args: Any, **kwargs: Any
+        self,
+        get_attribute: Callable[[T], Coroutine[Any, Any, R]],
+        instance: T,
+        name: str,
+        lock: AsyncContextManager[Any],
     ):
-        self.call = __call
-        self.args = args
-        self.kwargs = kwargs
+        self._get_attribute = get_attribute
+        self._instance = instance
+        self._name = name
+        self._lock = lock
+
+    def __await__(self) -> Generator[None, None, R]:
+        return self._await_impl().__await__()
 
-    def __await__(self) -> Generator[Any, Any, T]:
-        return self.call(*self.args, **self.kwargs).__await__()
+    async def _await_impl(self) -> R:
+        try:
+            stored = cast(Awaitable[R], self._instance.__dict__[self._name])
+        except KeyError:
+            # something deleted the cached value or future cached value placeholder. Restart
+            # the fetch by delegating to the cached_property descriptor.
+            return await cast(Awaitable[R], getattr(self._instance, self._name))
+
+        if stored is not self:
+            # another task produced a value
+            return await stored
+
```

Comment:
What dropping this will do is adding overhead to obtain a lock for any tasks that obtained the `_FutureCachedValue` instance from the `instance.__dict__` namespace before a cached result is produced.

E.g. consider this scenario:

1. Task 1 is obtains the `_FutureCachedValue` (== `fcv`) instance from the descriptor, which has also stored this on the `instance.__dict__` namespace.
2. Task 2 and 3 obtain the `fcv` instance from the `instance.__dict__` namespace.
3. Task 1 awaits on `fcv`, obtains the lock, and the getter coroutine completes to produce a value. This value is given to task 1, and stored in the `instance.__dict__` namespace, and the lock is released again.
4. Task 2 awaits on `fcv`, and without the fast path, attempts and succeeds in obtaining the lock. This uses `async with` so there is a potential task switch there.
5. Task 3 awaits on `fcv`, and without the fast path, is blocked on `async with self._lock`.
6. Task 2 is given the cached value (via `if stored is not self: return await stored` in the lock context).
7. Task 3 is unblocked and is given the cached value (via `if stored is not self: return await stored` in the lock context).

If `self._lock` is always `asyncio.Lock` this isn't a real issue because `asyncio.Lock().acquire()` returns immediately without further blocking if the lock is available. But that's not a guarantee for other async context managers, so potentially you could be introducing a lot of delay here if there are many tasks wanting to await on a shared `fcv`. This is not an uncommon scenario, because `instance.attribute_name` and `await fcv` can trivially be executed as two separate actions and real-life code _will_ do this (e.g. `tasks = [task_group.create_task(some_coroutine(instance.attribute) for _ in range(concurrency)]`).

Your scenario, where the `instance.__dict__` namespace no longer has _any_ object for the attribute name when awaiting on `fcv`, necessatitates the `try:...except KeyError:` exception handling blocks, but if that scenario wasn't possible, the many-tasks-with-fcv references scenario still exists.

So yes, you could drop the fast path here and avoid a little repetition, but I think that a library like asyncstdlib would not introduce too many potential roadblocks on performance here. :-)